File: ExelBidDemoObjC/ExelBidDemoObjC/StatusBadge.m
#import "StatusBadge.h"

@interface EBStatusBadge ()
@property (nonatomic, strong) UIView *dot;
@property (nonatomic, strong) UILabel *label;
@property (nonatomic, assign, readwrite) EBStatusBadgeState state;
@end

@implementation EBStatusBadge

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (!self) { return nil; }
    self.layer.cornerRadius = 14;
    self.layer.cornerCurve = kCACornerCurveContinuous;
    self.layer.masksToBounds = YES;

    _dot = [[UIView alloc] init];
    _dot.layer.cornerRadius = 4;
    _dot.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:_dot];

    _label = [[UILabel alloc] init];
    _label.font = [UIFont preferredFontForTextStyle:UIFontTextStyleFootnote];
    _label.numberOfLines = 0;
    _label.adjustsFontForContentSizeCategory = YES;
    _label.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:_label];

    [NSLayoutConstraint activateConstraints:@[
        [_dot.leadingAnchor  constraintEqualToAnchor:self.leadingAnchor constant:12],
        [_dot.centerYAnchor  constraintEqualToAnchor:self.centerYAnchor],
        [_dot.widthAnchor    constraintEqualToConstant:8],
        [_dot.heightAnchor   constraintEqualToConstant:8],
        [_label.leadingAnchor  constraintEqualToAnchor:_dot.trailingAnchor constant:8],
        [_label.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:-12],
        [_label.topAnchor      constraintEqualToAnchor:self.topAnchor constant:8],
        [_label.bottomAnchor   constraintEqualToAnchor:self.bottomAnchor constant:-8]
    ]];

    [self setState:EBStatusBadgeStateIdle text:@"Idle"];
    return self;
}

static UIColor *EBForegroundForState(EBStatusBadgeState state) {
    switch (state) {
        case EBStatusBadgeStateIdle:       return UIColor.secondaryLabelColor;
        case EBStatusBadgeStateLoading:    return UIColor.systemOrangeColor;
        case EBStatusBadgeStateReady:      return UIColor.systemBlueColor;
        case EBStatusBadgeStateImpression: return UIColor.systemGreenColor;
        case EBStatusBadgeStateClicked:    return UIColor.systemPurpleColor;
        case EBStatusBadgeStateFailed:     return UIColor.systemRedColor;
    }
}

- (void)setState:(EBStatusBadgeState)state text:(NSString *)text {
    self.state = state;
    UIColor *fg = EBForegroundForState(state);
    self.label.text = text;
    self.label.textColor = fg;
    self.dot.backgroundColor = fg;
    self.backgroundColor = [fg colorWithAlphaComponent:0.12];
}

@end

File: ExelBidDemo/ExelBidDemo/StatusBadge.swift
import UIKit

/// Pill-shaped status indicator for the ad lifecycle. Tinted background +
/// matching dot so the same state reads identically across screens.
final class StatusBadge: UIView {

    enum State {
        case idle, loading, ready, impression, clicked, failed
    }

    private let dot = UIView()
    private let label = UILabel()
    private(set) var state: State = .idle

    override init(frame: CGRect) {
        super.init(frame: frame)
        layer.cornerRadius = 14
        layer.cornerCurve = .continuous
        layer.masksToBounds = true

        dot.layer.cornerRadius = 4
        dot.translatesAutoresizingMaskIntoConstraints = false
        addSubview(dot)

        label.font = .preferredFont(forTextStyle: .footnote)
        label.numberOfLines = 0
        label.adjustsFontForContentSizeCategory = true
        label.translatesAutoresizingMaskIntoConstraints = false
        addSubview(label)

        NSLayoutConstraint.activate([
            dot.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 12),
            dot.centerYAnchor.constraint(equalTo: centerYAnchor),
            dot.widthAnchor.constraint(equalToConstant: 8),
            dot.heightAnchor.constraint(equalToConstant: 8),

            label.leadingAnchor.constraint(equalTo: dot.trailingAnchor, constant: 8),
            label.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -12),
            label.topAnchor.constraint(equalTo: topAnchor, constant: 8),
            label.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -8)
        ])

        set(.idle, text: "Idle")
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError() }

    func set(_ state: State, text: String) {
        self.state = state
        label.text = text
        label.textColor = state.foreground
        dot.backgroundColor = state.foreground
        backgroundColor = state.background
    }
}

private extension StatusBadge.State {
    var foreground: UIColor {
        switch self {
        case .idle:       return .secondaryLabel
        case .loading:    return .systemOrange
        case .ready:      return .systemBlue
        case .impression: return .systemGreen
        case .clicked:    return .systemPurple
        case .failed:     return .systemRed
        }
    }
    var background: UIColor {
        foreground.withAlphaComponent(0.12)
    }
}
